ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -106,13 +107,15 @@ public OperationController(String suffix, String defaultPartitionClass, AccountS
     this.responseHandler = responseHandler;
     this.nonBlockingRouter = nonBlockingRouter;
     // Warm up connections to dataNodes in local and remote DCs.
-    List<ResponseInfo> responseInfos = Collections.synchronizedList(new ArrayList<>());
     String localDatacenter = clusterMap.getDatacenterName(clusterMap.getLocalDatacenterId());
     Map<Boolean, List<DataNodeId>> localAndRemoteNodes = clusterMap.getDataNodeIds()
         .stream()
         // ignore any in-process "data nodes" without TCP ports
         .filter(dataNodeId -> dataNodeId.getPort() != DataNodeId.UNKNOWN_PORT)
         .collect(Collectors.partitioningBy(dataNodeId -> localDatacenter.equals(dataNodeId.getDatacenterName())));
+    // 'responseInfos' list needs to be thread safe since it it possible that the list can be modified as we are reading
+    // it for scenario where we returned from networkClient.warmUpConnections() due to time out.
+    List<ResponseInfo> responseInfos = new CopyOnWriteArrayList<>();
```

Comment:
COW array is not a good idea. It just creates a new array on each write and consumes more memory. 
Does synchronizedList not guard the array using a lock ?

Reply:
nvm, i found that syncList is not thread-safe when iterating and a concurrent write (!). COW is ok iff the list is small and there are a few writes.